src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/TypeFixupSignature.cs

```diff
@@ -39,11 +42,114 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
                 EcmaModule targetModule = factory.SignatureContext.GetTargetModule(_typeDesc);
                 SignatureContext innerContext = dataBuilder.EmitFixup(factory, _fixupKind, targetModule, factory.SignatureContext);
                 dataBuilder.EmitTypeSignature(_typeDesc, innerContext);
+
+                if (_fixupKind == ReadyToRunFixupKind.Check_TypeLayout)
+                {
+                    EncodeTypeLayout(dataBuilder, _typeDesc);
+                }
             }
 
             return dataBuilder.ToObjectData();
         }
 
+        private static void EncodeTypeLayout(ObjectDataSignatureBuilder dataBuilder, TypeDesc type)
+        {
+            Debug.Assert(type.IsValueType);
+            MetadataType defType = (MetadataType)type;
+
+            int pointerSize = type.Context.Target.PointerSize;
+            int size = defType.InstanceFieldSize.AsInt;
+            int alignment = GetClassAlignmentRequirement(defType);
+            ReadyToRunTypeLayoutFlags flags = ReadyToRunTypeLayoutFlags.READYTORUN_LAYOUT_Alignment | ReadyToRunTypeLayoutFlags.READYTORUN_LAYOUT_GCLayout;
+            if (alignment == pointerSize)
+            {
+                flags |= ReadyToRunTypeLayoutFlags.READYTORUN_LAYOUT_Alignment_Native;
+            }
+
+            if (!defType.ContainsGCPointers)
+            {
+                flags |= ReadyToRunTypeLayoutFlags.READYTORUN_LAYOUT_GCLayout_Empty;
+            }
+
+            bool isHfa = defType.IsHfa && defType.Context.Target.Architecture == TargetArchitecture.ARM || defType.Context.Target.Architecture == TargetArchitecture.ARM64;
```

Comment:
The IsHFA computation in crossgen2 looks slightly different from what CoreCLR does.

E.g. we might need to change the implementation of `ComputeValueTypeShapeCharacteristics` on `VectorFieldLayoutAlgorithm` to account for their special HFA status.

Reply:
(But it looks like a pre-existing problem that we should just track somewhere)

Reply:
> The IsHFA computation in crossgen2 looks slightly different from what CoreCLR does.
> 
> E.g. we might need to change the implementation of `ComputeValueTypeShapeCharacteristics` on `VectorFieldLayoutAlgorithm` to account for their special HFA status.

I have implemented that. Need to merge with this PR.